File: src/hack_forked/utils/sys/pidLog.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Hh_core

let log_oc = ref None

let enabled = ref true

let disable () = enabled := false

let init pids_file =
  assert (!log_oc = None);
  Sys_utils.with_umask 0o111 (fun () ->
      Sys_utils.mkdir_no_fail (Filename.dirname pids_file);
      let oc = open_out pids_file in
      log_oc := Some oc;
      Unix.(set_close_on_exec (descr_of_out_channel oc)))

let log ?reason ?(no_fail = false) pid =
  if !enabled then
    let pid = Sys_utils.pid_of_handle pid in
    let reason =
      match reason with
      | None -> "unknown"
      | Some s -> s
    in
    match !log_oc with
    | None when no_fail -> ()
    | None -> failwith "Can't write pid to uninitialized pids log"
    | Some oc -> Printf.fprintf oc "%d\t%s\n%!" pid reason

exception FailedToGetPids

let get_pids pids_file =
  try
    let ic = open_in pids_file in
    let results = ref [] in
    begin
      try
        while true do
          let row = input_line ic in
          if Str.string_match (Str.regexp "^\\([0-9]+\\)\t\\(.+\\)") row 0 then
            let pid = int_of_string (Str.matched_group 1 row) in
            let reason = Str.matched_group 2 row in
            results := (pid, reason) :: !results
        done
      with
      | End_of_file -> ()
    end;
    close_in ic;
    List.rev !results
  with
  | Sys_error _ -> raise FailedToGetPids

let close () =
  Base.Option.iter !log_oc ~f:close_out;
  log_oc := None

File: rust_port/crates/flow_daemon/src/pid_log.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fs;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;
use std::io::Write;
use std::path::Path;
use std::sync::Mutex;
use std::sync::OnceLock;

use flow_common::sys_utils;

fn log_oc() -> &'static Mutex<Option<File>> {
    static LOG_OC: OnceLock<Mutex<Option<File>>> = OnceLock::new();
    LOG_OC.get_or_init(|| Mutex::new(None))
}

fn enabled() -> &'static Mutex<bool> {
    static ENABLED: OnceLock<Mutex<bool>> = OnceLock::new();
    ENABLED.get_or_init(|| Mutex::new(true))
}

pub fn disable() {
    *enabled().lock().expect("pid_log enabled mutex poisoned") = false;
}

pub fn init(pids_file: &Path) -> std::io::Result<()> {
    let mut guard = log_oc().lock().expect("pid_log log_oc mutex poisoned");
    assert!(guard.is_none(), "PidLog::init called twice");
    sys_utils::with_umask(0o111, || -> std::io::Result<()> {
        if let Some(parent) = pids_file.parent() {
            sys_utils::mkdir_no_fail(parent)?;
        }
        let oc = fs::OpenOptions::new()
            .create(true)
            .truncate(true)
            .write(true)
            .open(pids_file)?;
        *guard = Some(oc);
        Ok(())
    })
}

pub fn log(reason: Option<&str>, no_fail: bool, pid: u32) {
    if !*enabled().lock().expect("pid_log enabled mutex poisoned") {
        return;
    }
    let pid = sys_utils::pid_of_handle(pid);
    let reason = reason.unwrap_or("unknown");
    let mut guard = log_oc().lock().expect("pid_log log_oc mutex poisoned");
    match guard.as_mut() {
        None if no_fail => {}
        None => panic!("Can't write pid to uninitialized pids log"),
        Some(oc) => {
            writeln!(oc, "{}\t{}", pid, reason).expect("PidLog::log: failed to write");
            oc.flush().expect("PidLog::log: failed to flush");
        }
    }
}

#[derive(Debug)]
pub struct FailedToGetPids;

impl std::fmt::Display for FailedToGetPids {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "FailedToGetPids")
    }
}

impl std::error::Error for FailedToGetPids {}

pub fn get_pids(pids_file: &Path) -> Result<Vec<(u32, String)>, FailedToGetPids> {
    let ic = File::open(pids_file).map_err(|_| FailedToGetPids)?;
    let reader = BufReader::new(ic);
    let mut results: Vec<(u32, String)> = Vec::new();
    for line in reader.lines() {
        let row = match line {
            Ok(row) => row,
            Err(_) => break,
        };
        let mut parts = row.splitn(2, '\t');
        let Some(pid_str) = parts.next() else {
            continue;
        };
        let Some(reason) = parts.next() else {
            continue;
        };
        let Ok(pid) = pid_str.parse::<u32>() else {
            continue;
        };
        results.push((pid, reason.to_string()));
    }
    Ok(results)
}

pub fn close() -> std::io::Result<()> {
    let mut guard = log_oc().lock().expect("pid_log log_oc mutex poisoned");
    if let Some(mut oc) = guard.take() {
        oc.flush()?;
        drop(oc);
    }
    Ok(())
}
